### ISSUE TITLE
Issue with Weight Synchronization When Using Consecutive DDP Instances in the Same torch.distributed Environment

### ISSUE THREAD

**Issue by joansaurina** [pytorch/pytorch#134960]
### 🐛 Describe the bug

Hello, I am working on a project where I need to use multiple consecutive instances of DistributedDataParallel (DDP) within the same torch.distributed environment. In my scenario, I initially train a model, then freeze some layers and create a second model by reusing a layer from the first one. 

However, I am observing that when using the second DDP instance for the second model, the weights are not synchronizing correctly across multiple GPUs. While the first model seems to work fine, the second model shows discrepancies in the weights between GPUs. 

Is it possible to have multiple consecutive DDP instances within the same torch.distributed environment? Or are there any special considerations I need to be aware of to ensure the weights are synchronized correctly? I would appreciate any guidance or recommendations on this issue.

Thank you.

### Versions

-

**Comment by wz337** [comment]
cc. @fegin

**Comment by fegin** [comment]
I'm not sure if I fully understand the question. Are you training two models simultaneously and these two models share some weights? In such a case, I can imagine the gradients of the shared weights may not synchronized correctly by the two different DDP instances. One workaround is to use a model wrapper to wrap the two models and use only one DDP instance to wrap the model wrapper.
